### PR TITLE
feat(rpc-eth-types+api): add latest chain state tracking to `EthStateCache ` and `EthStateCacheService`

### LINKED ISSUE #13069: Track latest block in rpc cache
### Describe the feature

we have a race condition that affects the `Latest` block:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-api/src/helpers/block.rs#L242-L254

by the time the cache handles this command:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-types/src/cache/mod.rs#L499-L502

the latest block can be reorged in between and removed from the cache:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-types/src/cache/mod.rs#L526-L528

we can get rid of this race condition, in a few ways:
1. retry on the callsite if it returns None and the blockId is latest
2. track the latest block in the cache itself: last entry in:
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-types/src/cache/mod.rs#L524-L524

1. would be easier but feels hacky

2. is cleaner because this approach then allows us to also track the safe+finalized in a similar way.

## TODO
* add an optional field to to statecache that tracks the last received `ChainChange` then add a `latest_blockwith_senders` function to ethstatecache

### Additional context

https://github.com/paradigmxyz/reth/issues/13001

### PR DESCRIPTION
**Pull request by lean-apple**
Closes https://github.com/paradigmxyz/reth/issues/13069. 

Only use  `latest_block_with_senders ` in `crates/rpc/rpc-eth-api/src/helpers/block.rs` for now.

### CODE REVIEW

File: crates/rpc/rpc-eth-api/src/helpers/block.rs [hunk 1]
```diff
@@ -253,19 +262,48 @@ pub trait LoadBlock: LoadPendingBlock + SpawnBlocking + RpcNodeCoreExt {
...
+                }
+                _ => {
+                    let block_hash = match self
+                        .provider()
+                        .block_hash_for_id(block_id)
+                        .map_err(Self::Error::from_eth_err)?
+                    {
+                        Some(block_hash) => block_hash,
+                        None => return Ok(None),
+                    };
```
Comment (lean-apple) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:298]: @mattsse A bit stuck on the integration of the new function `latest_block_with_senders`, I have to add this fallback to fix the integration test but the code is very redundant

File: crates/rpc/rpc-eth-api/src/helpers/block.rs [hunk 2]
```diff
@@ -253,19 +262,48 @@ pub trait LoadBlock: LoadPendingBlock + SpawnBlocking + RpcNodeCoreExt {
...
-            let block_hash = match self
-                .provider()
-                .block_hash_for_id(block_id)
-                .map_err(Self::Error::from_eth_err)?
-            {
-                Some(block_hash) => block_hash,
-                None => return Ok(None),
-            };
+            let block = match block_id {
+                BlockId::Number(BlockNumberOrTag::Latest) => {
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:266]: I think we can use `is_latest()` here

File: crates/rpc/rpc-eth-types/src/cache/mod.rs [hunk 3]
```diff
@@ -61,15 +61,10 @@ type HeaderLruCache<H, L> = MultiConsumerLruCache<B256, H, L, HeaderResponseSend
...
+#[derive(Debug, Clone)]
 pub struct EthStateCache<B: Block, R> {
     to_service: UnboundedSender<CacheAction<B, R>>,
-}
-
-impl<B: Block, R> Clone for EthStateCache<B, R> {
-    fn clone(&self) -> Self {
-        Self { to_service: self.to_service.clone() }
-    }
+    latest_chain_change: Option<ChainChange<B, R>>,
```
Comment (mattsse) [inline crates/rpc/rpc-eth-types/src/cache/mod.rs:67]: we don't need to share this and can instead rely on message passing and always fetch this from the service

File: crates/rpc/rpc-eth-types/src/cache/mod.rs [hunk 4]
```diff
@@ -186,6 +182,21 @@ impl<B: Block, R: Send + Sync> EthStateCache<B, R> {
...
+    /// reorgs.
+    /// Returns `None` if no canonical chain is tracked or during reorgs.
+    pub async fn latest_block_with_senders(
+        &self,
+    ) -> ProviderResult<Option<Arc<SealedBlockWithSenders<B>>>> {
+        if let Some(chain_change) = &self.latest_chain_change {
+            if let Some(latest_block) = chain_change.blocks.last() {
+                return self.get_sealed_block_with_senders(latest_block.hash()).await;
+            }
+        }
```
Comment (mattsse) [inline crates/rpc/rpc-eth-types/src/cache/mod.rs:197]: this should always send an action to the service, but for fetching the latest, we'd need a new action variant

**Review by mattsse — CHANGES_REQUESTED** [review]
the last entry in this list will always be the `latest` block

https://github.com/paradigmxyz/reth/blob/9141db79c9233dbc59b12527b967a8061ee7e66a/crates/rpc/rpc-eth-types/src/cache/mod.rs#L460-L463

so we can track that separately.

and to fetch the latest block, we can add a new action variant, like this but without the hash:

https://github.com/paradigmxyz/reth/blob/9141db79c9233dbc59b12527b967a8061ee7e66a/crates/rpc/rpc-eth-types/src/cache/mod.rs#L498-L501

File: crates/rpc/rpc-eth-api/src/helpers/block.rs (hunk 1 above)
Comment (mattsse) [reply crates/rpc/rpc-eth-api/src/helpers/block.rs:298]: we can look at this later, I believe we can simplify this a bit by introducing a tmp var `let mut maybe_block = None;`

and then do a `if block_id.is_latest() { maybe_block = ... }`

**Pull request closed without merging**